### ISSUE TITLE
Switch to CMake (was: Native build on Windows using Clang)

### ISSUE THREAD

**Issue by DesWurstes** [bitcoin/bitcoin#14118]
Bitcoin Core can be built on Windows using Clang natively. Currently only a few projects are using this (Chrome [since march](http://blog.llvm.org/2018/03/clang-is-now-used-to-build-chrome-for.html), Firefox [since yesterday](https://news.ycombinator.com/item?id=17504197)). It'd make building for Window easier. Moreover, I'm sure that it's easier to use than MSYS + MinGW.

There's only one obstacle:

- Autotools does not run on Windows!

What are your opinions on creating a Cmake file? Although the `configure.ac` is not-so-tiny, the review community is really active here. If I get positive feedback about it, I'd be glad to create the first draft of it. In addition, Cmake files are used more often. Wouldn't it be great to give the option for builders to choose between autotools and Cmake?

**Comment by ken2812221** [comment]
Concept ACK.
This should be very helpful for those who want to build on Windows.

**Comment by jenishhhh** [comment]
who want to build on Windows.google.com

**Comment by scravy** [comment]
When moving to cmake alltogether there would be only one build system which is supported by a zillion IDEs and works on all operating systems.

**Comment by sipa** [comment]
On Sun, Sep 2, 2018, 04:32 Julian Fleischer <notifications@github.com>
wrote:

> When moving to cmake alltogether there would be only one build system
> which is supported by a zillion IDEs and works on all operating systems.
>
I'm sure there are advantages of one build system over another, and if
cmake can replace all supported existing use cases and more, then that
sounds like a good deal.

However, it does sound like a monumental task to me. I don't know much
about build systems, so maybe I'm overestimating this, but does this mean
we also need to cmakeify libsecp256k1 and libunivalue? Does the depends
build need to be modified as well, or can we keep that make based? Do we
need to find alternatives to the make based tricks used to run the unit
tests in parallel? I'm just concerned that people suggesting moving from
one setup to another underestimate the work that went into our current
build infrastructure.

>

**Comment by ken2812221** [comment]
IMO, we are already maintain 2 different build system(MSVC) which integrate by a weird way (auto-generated by python script). Switching to cmake isn't going to add the third build system but to replace native MSVC project files to be more readable and easier to edit. If the cmake build system have done, the msvc native *.vcxproj files could be removed immediately. Also, there are more and more projects are switching to cmake because it's more portable, especially for MSVC and clang-cl support.

**Comment by DesWurstes** [comment]
I've opened a UniValue Pull Request: https://github.com/bitcoin-core/univalue/pull/16

will work on Core Cmake soon

> but does this mean we also need to cmakeify libsecp256k1 and libunivalue?

Cmake can use Autoconf dependencies; that's needed for Windows compatibility

> Does the depends build need to be modified as well, or can we keep that make based?

Windows has `nmake`, so that's not needed, although Cmake can download dependencies.

> Do we need to find alternatives to the make based tricks used to run the unit tests in parallel?

no idea

**Comment by jameshilliard** [comment]
I'd generally recommend against moving away from autotools for this type of project, while autotools syntax is rather archaic/terrible it does do a number of things well that other build systems do not, mostly when it comes to extreme flexibility, portability in addition to properly handling cross-compiliation. We need to be able to cross compile binaries for all targets that we do binary releases for from a linux build system due to gitian. For developers on Windows we do have [msvc build](https://github.com/bitcoin/bitcoin/tree/master/build_msvc) support already.

One issue with cmake is that it's a makefile generator while autoconf is a configure script generator, the configure script created by autoconf is designed to be portable, there isn't really a direct equivalent with cmake AFAIK. In general autoconf projects are considered easier to work with for people packaging software even though the build system may be somewhat more complex to maintain.

**Comment by Talkless** [comment]
> Do we need to find alternatives to the make based tricks used to run the unit tests in parallel? 

It is possible to parallelize tests using `ctest` [0]. CMake Cookbook example [1] shows that it is possible to group (if needed) tests with costs attached.

> We need to be able to cross compile binaries for all targets that we do binary releases for from a linux build system due to gitian.

Are you implying that CMake does not support cross-compilation..? You "just" (heh) define toolchain file. Same CMake Cookbook has example where you build Windows executable from Linux host. See docs [2].

> For developers on Windows we do have msvc build support already.

If CMake would generate MSVC projects for us, that would mean maintaining that directory would be no longer needed?

[0] https://cmake.org/cmake/help/v3.14/manual/ctest.1.html
[1] https://github.com/dev-cafe/cmake-cookbook/blob/v1.0/chapter-04/recipe-08/using-cost-example/CMakeLists.txt
[2] https://cmake.org/cmake/help/v3.14/manual/cmake-toolchains.7.html#cross-compiling

**Comment by sipa** [comment]
> Are you implying that CMake does not support cross-compilation..?

@Talkless No, I'm implying anything; I'm listing perhaps nontrivial things that we've built based on our existing build system, which would need to be redone for a new build system before we can even consider switching.

**Comment by jameshilliard** [comment]
> Are you implying that CMake does not support cross-compilation..? 

No, but I find autotools to typically have much better cross compilation support than CMake.

**Comment by elichai** [comment]
Is anyone working on this?

I think trying to switch to a simpler build system is a good idea (assuming the result is actually simpler)
And cmake is now a requirement for some stuff(CLion in particular)

**Comment by MarcoFalke** [comment]
@elichai Not that I am aware of. Also @theuni seems to be working on keeping the dependencies in autotools. IIUC,  as a first step they would have to be transitioned to cmake.

See the pull from @theuni 
*  Split sources for easier buildsystem integration https://github.com/bitcoin-core/univalue/pull/19

**Comment by MarcoFalke** [comment]
>  And cmake is now a requirement for some stuff(CLion in particular)

You can edit and run Bitcoin Core in Clion, but Clion can't do the static code analysis.

**Comment by elichai** [comment]
Thanks.
I think I succeeded to "hack" CLion to recognize all the files by just adding a dummy CMakeLists file (https://gist.github.com/elichai/60f1a16fdb51aeba9a9de0a5de23738a)

looking at @theuni PR it seems like working on CMake is a waste of time (unless someone can really show that the result will be a lot simpler to manage and understand)

**Comment by chiumichael** [comment]
I think it's worth adding CMake support given the broader c++ community has basically settled on CMake (even Qt6 is switching to it in favor of qmake). It would take quite a bit of time and happen incrementally, but there's nothing wrong with that and it would allow others not familiar with CMake to understand it better; I'd be more than glad to work on it. 

P.S. On a related note, why are some of the dependencies like libsecp256k1 copy and pasted in and not pulled as a submodule?

**Comment by elichai** [comment]
@choumichael it's not copy-pasted.
It's a git subtree
https://manpages.debian.org/testing/git-man/git-subtree.1.en.html

**Comment by jameshilliard** [comment]
If anything [meson](https://mesonbuild.com/) is probably a better choice than cmake by now and has [visual studio support](https://mesonbuild.com/Using-with-Visual-Studio.html). A number of open source projects such as gnome have been [porting to meson from autotools](https://wiki.gnome.org/Initiatives/GnomeGoals/MesonPorting).

**Comment by chiumichael** [comment]
[build2](https://build2.org/faq.xhtml) is interesting too imo

**Comment by MarcoFalke** [comment]
It appears, meson could also be wrapped into cmake IDEs (https://pypi.org/project/meson-cmake-wrapper/)

**Comment by elichai** [comment]
Sadly if people can't even agree on a specific "best" build system then the chances of this ever being accepted is pretty low.

I would personally be in favor of switching a build system *only* if it gives us one of these features:

1. Reduce the complexity of the build system by a lot (making it easier to maintain and adjust over time)
2. Reduce the amount of stuff to maintain (i.e. MSVC stuff).
3. Can show actual real superiority over the current build system (highly doubt that can be the case).

A nice feature but I don't think that's enough to switch is somehow faster builds.

**Comment by sipsorcery** [comment]
@fanquake @MarcoFalke imho this issue should be closed pending someone coming up with a cmake PR.

The OP mentions autotools and msys + mingw there is unlikely to be any consideration given to switching the official build system to cmake. Most of the discussion is about switching the current msvc build (which is for developer convenience only and never used to build official binaries) to cmake. That would be possibly be a minor improvement to the current msvc build config and Visual Studio does now have [cmake support](https://docs.microsoft.com/en-us/cpp/build/cmake-projects-in-visual-studio?view=vs-2019). Without a PR, even it was for only one of the libraries, this issue is a bit academic.

**Comment by MarcoFalke** [comment]
The feature request didn't seem to attract much attention in the past. Also, the issue seems not important enough right now to keep it sitting around idle in the list of open issues.

Closing due to lack of interest. Pull requests with improvements are always welcome.

**Comment by hebasto** [comment]
> Bitcoin Core can be built on Windows using Clang natively.

Please see https://github.com/bitcoin/bitcoin/pull/31507.
